### ISSUE TITLE
How to control the thread number on Tensorflow?

### ISSUE THREAD

**Issue by Ellie68** [tensorflow/tensorflow#14900]
Hi all,

Do we have any options to control the number of threads in TF-Slim both in training and evaluation processes?

Specifically, I use [this network](https://github.com/pudae/tensorflow-densenet) for my classification problem. I changed the evaluation part in a way that runs train and evaluation in parallel like [this code](https://github.com/mnuke/tf-slim-mnist). I can run it on my own CPU without any problem. But I can't execute them on a supercomputer. It seems that it is related to the very large number of threads which are being created by Tensorflow. If the number of threads exceeds the maximum number of threads pre-set in SLURM (= 28) then the job will fail. Since it's unable to create new threads it will end up with error "resource temporarily unavailable".

This error provided when the code tries to restore parameters from checkpoints. If there is no limitation on the number of threads (like on my pc) it works fine:
```
INFO:tensorflow:Restoring parameters from ./model.ckpt-0
INFO:tensorflow:Starting evaluation at
I tensorflow/core/kernels/logging_ops.cc:79] eval/Accuracy[0]
I tensorflow/core/kernels/logging_ops.cc:79] eval/Recall_5[0]
INFO:tensorflow:Evaluation [1/60]
```

However, when there is a limitation on the number of threads (like SLURM job submission on supercomputers) we get:
```
INFO:tensorflow:Restoring parameters from ./model.ckpt-0
terminate called after throwing an instance of 'std::system_error'
  what():  Resource temporarily unavailable
```

I tried to limit the number of CPU threads used by Tensorflow to 1 by creating config like:

     FLAGS.num_preprocessing_threads=1
      config = tf.ConfigProto()
      config.intra_op_parallelism_threads = FLAGS.num_preprocessing_threads
      config.inter_op_parallelism_threads = FLAGS.num_preprocessing_threads
    
        slim.evaluation.evaluation_loop(
            master=FLAGS.master,
            checkpoint_path=each_ckpt,
            logdir=FLAGS.eval_dir,
            num_evals=num_batches,
            eval_op=list(names_to_updates.values()) + print_ops,
            variables_to_restore=variables_to_restore,
            session_config=config)
But unfortunately, that didn't help. In my opinion, the main problem we are having here is the fact that we are not able to control the number of threads here. Although we set it to 1 with various TF options you can actually see that this job is creating many more threads on the node:

slurm_script─┬─python───128*[{python}]
             └─python───8*[{python}]

Training script is creating 128 threads and evaluation script is creating 8 (both numbers vary over time).  

Any idea on the way to control the thread numbers will be highly appreciated because I do need to fix this issue urgently.
Ellie

P.S. I'm using Python 2.7.13 and Tensorflow 1.3.0.

**Comment by aselle** [comment]
@mrry, do you have any ideas on this? I'd imagine python creates many threads, but it's strange that slurm actually limits the number of threads. A lot of tensorflow threads are IO threads that don't really correspond to an allocation of CPU resources. Perhaps you can talk to your SLURM sysadmin to increase this limit.

**Comment by mrry** [comment]
The main additional source of threads would be from the Slim input pipeline, which uses Python queue runners. From your code fragment I can't tell how many threads would be created, but you might be able to set a smaller number of threads when creating the pipeline.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue? Please update the label and/or status accordingly.

**Comment by Ellie68** [comment]
Hi @aselle 
Unfortunately, SLURM sysadmin can't increase the limits for me.

**Comment by Ellie68** [comment]
Hi @mrry 
Thanks for your answer. I mentioned the way that I tried to set a smaller number of threads, but it doesn't help.
Do you have any suggestion for me to do this?
As @pudae mentioned [here](https://github.com/pudae/tensorflow-densenet/issues/7#issuecomment-351054685) 
"It seems that there is no way to reduce number of threads with using slim.
I tested evaluation scripts with num_preprocessing_threads=1,
then I found that slim create 6 threads for input pipeline and 1 thread for event logger."

> 

>

**Comment by mrry** [comment]
You could try switching your code to use `tf.data` for the input pipeline and `tf.train.MonitoredSession` for the session. These classes create far fewer threads than the Slim-based training code, and should be able to fit in the limit of 28 threads if you also limit the `inter_op_parallelism_threads` and `intra_op_parallelism_threads`.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue? Please update the label and/or status accordingly.

**Comment by mrry** [comment]
Closing this due to lack of activity. Hopefully it worked and you're happily training with a smaller number of threads!
